docs/HARDWARE-MODEL.md

```diff
@@ -185,6 +185,93 @@ In this case it would make sense to try with:
     }
 ```
 
+## Container Device Interface support for GPUs
+
+Some devices, specially those integrated to a System on a Chip (SoC), might require complex setup in order to be exposed
+to Edge Applications. Additionally, if the device is not connected to a PCIe bus and/or the SoC doesn't implement an
+IOMMU (or SMMU on ARM), it cannot be exposed (passthrough) to Edge Applications using the VFIO mechanism. For example,
+the GPU present on NVIDIA Jetson devices is not connected through the PCIe bus, so it cannot be exposed to any Virtual
+Machine.
+
+For such specific cases, EVE-OS supports the Container Device Interface (CDI) mechanism that can be used to exposed a
+GPU device to a bare metal container. It operates as follows:
+
+1. A CDI (yaml) file is provided at /etc/cdi on the rootfs. This file contains all the specific setup to expose a GPU to
+   bare metal containers, such as access to file devices, host libraries that needs to be mounted inside the container,
+   custom actions to be performed, like creation of symbolic links, etc.
+1. Each setup is identified by a name in the format "vendor.com/device" inside the CDI file. For instance, a device with
+   two integrated GPUs might have defined "vendor.com/gpu=0" and "vendor.com/gpu=1".
+1. A HDMI I/O adapter must be specified in the device model, configuring the CDI string using the "cbattr" attribute.
+1. Some GPUs can be accessed by multiple containers during runtime, for such cases, a single HDMI I/O adapter should be
+   created per container.
+1. The Edge Application (bare metal container) is deployed with the corresponding I/O adapter attached to the container.
+
+A HDMI I/O adapter (ztype 7) with CDI support can be defined as the following:
+
+```json
+{
+    "ztype": 7,
+    "phylabel": "GPU0",
+    "phyaddrs": {},
+    "logicallabel": "GPU0",
+    "assigngrp": "group1",
+    "usage": "ADAPTER_USAGE_UNSPECIFIED",
+    "cbattr": {
+        "cdi": "vendor.com/gpu=0"
+    },
+    "usagePolicy": {},
+    "cost": 0,
+    "vfs": null,
+    "parentassigngrp": ""
+}
+```
+
+If two containers need to access the same GPU (when supported), two I/O adapters must be defined. For instance, two
```

Comment:
I don't think so, but will let Rene confirm. You have one or more CDI yaml files that describe the devices on the device and how to access them (including devices, and mounts necessary and all sorts of instruction stuff). Whether it is 1 or 50 apps, only the defined CDI files for the device are used. They are read-only.

Reply:
Exactly @deitch ... @OhmSpectator , we must have one GPU I/O adapter per Edge App because an I/O adapter can be "passed-through" to only one VM, makes total sense for PCI devices, but we do have some exceptions, like CAN interfaces (that can be accessed by multiple VMs at the same time) and now this use case, where a single GPU can be shared across multiple containers... but this is only about device model, there is nothing to do with CDI files.....